### ISSUE TITLE
DynamicProperty lifecycle behaviour

### ISSUE THREAD

**Issue by Gurpartap** [ReactiveCocoa/ReactiveCocoa#2068]
###### A)

In my usage, as below, I had a few MutableProperty instances bound to DynamicProperty.

```Swift
func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
	let cell = tableView.dequeueReusableCellWithIdentifier(viewModel.cellIdentifier, forIndexPath: indexPath) as! PlaylistCellView

	let cellViewModel = viewModel.cellViewModels.value[indexPath.row]

	let reuseSignal = prepareForReuseSignal(cell)
	cellViewModel.observeTrack(until: reuseSignal)

	DynamicProperty(object: cell.progressView, keyPath: "hidden") <~ cellViewModel.hideProgressView.producer |> takeUntil(reuseSignal) |> map { $0 }
	DynamicProperty(object: cell.progressView, keyPath: "progress") <~ cellViewModel.downloadProgress.producer |> takeUntil(reuseSignal) |> map { $0 }

	// ...

	return cell
}
```

However, #1857 set it such that DynamicProperty of `cell.progressView` will be retained as long as `cell.progressView` is not deallocated ([Property.swift#L170](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/c3f9b4729c078d395e638819bdfc8889b9c9d956/ReactiveCocoa/Swift/Property.swift#L170)).

I can see how this enables ad-hoc DynamicProperty().producer implementations. However, doesn't this make the above use case of `DynamicProperty with <~` basically ill-fated? I expected to be able to see DynamicProperty releasing when the MutableProperty's producer terminates or when `reuseSignal` is invoked. The end result here is that the table view's performance gradually degrades the more you scroll it up or down.

###### B)

OK. Assuming that I'm not supposed to use it the ill-fated way, instead of creating a new DynamicProperty every time `tableView: cellForRowAtIndexPath:` is called, I moved the initialization to `UITableViewCell` implementation as such:

```Swift
class PlaylistCellView: UITableViewCell {

	@IBOutlet var progressView: XYZProgressView!

	lazy var progressViewHidden: DynamicProperty = {
		return DynamicProperty(object: self.progressView, keyPath: "hidden")
	}()

	lazy var progressViewProgress: DynamicProperty = {
		return DynamicProperty(object: self.progressView, keyPath: "progress")
	}()

	// ...
}
```

And changed the implementation in `tableView: cellForRowAtIndexPath:` as such:

```Swift
func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
	// ...
	cell.progressViewHidden <~ cellViewModel.hideProgressView.producer |> takeUntil(reuseSignal) |> map { $0 }
	cell.progressViewProgress <~ cellViewModel.downloadProgress.producer |> takeUntil(reuseSignal) |> map { $0 }
	// ...
}
```

I expected this workaround to counter my findings in **A**. However, it does not seem to help with the gradually degrading performance as you scroll around the table view. I haven't been able to figure out why yet.

###### C)

On the other hand, this simple implementation makes the table view scrolling butter smooth:

```Swift
func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
	// ...
	cellViewModel.hideProgressView.producer |> takeUntil(reuseSignal) |> start(next: { [weak cell] hidden in
		cell?.progressView.hidden = hidden
	})

	cellViewModel.downloadProgress.producer |> takeUntil(reuseSignal) |> start(next: { [weak cell] progress in
		cell?.progressView.progress = progress
	})
	// ...
}
```

TLDR:

A) Using `DynamicProperty() <~ MutableProperty().producer` in `tableView: cellForRowAtIndexPath:` comes with the caveat that it cannot dealloc DynamicProperty as the cell is reused. Scrolling performance degrades as the table view scrolls.

B) Moving DynamicProperty() to UITableViewCell subclass (so that it is initialized only once per cell) does not seem to help with the performance either, as I thought it would. There's, perhaps, something I haven't realized well here.

C) Listening to MutableProperty().producer and forwarding the value directly works as expect without any performance degradation.

**Comment by neilpa** [comment]
I recommend setting up the bindings once when initializing the cell subclass instead of each time a cell is requested. Generally this requires exposing a `MutableProperty` for your view model on a that can be `nil`. Then internally the cell binds the view model property to `DynamicProperty`s for the pieces of UI that need updating.

**Comment by ikesyo** [comment]
Closing this due to inactivity.
